packages/mui-lab/src/Masonry/Masonry.js

```diff
@@ -23,46 +22,115 @@ const useUtilityClasses = (ownerState) => {
   return composeClasses(slots, getMasonryUtilityClass, classes);
 };
 
+const computeBreakpointsBase = (breakpoints, prop) => {
+  const base = {};
+  if (Array.isArray(prop)) {
+    const arrayLen = prop.length;
+    Object.keys(breakpoints.values).forEach((breakpoint, i) => {
+      if (i < arrayLen) {
+        base[breakpoint] = true;
+      }
+    });
+  } else {
+    Object.keys(breakpoints.values).forEach((breakpoint) => {
+      if (prop[breakpoint] != null) {
+        base[breakpoint] = true;
+      }
+    });
+  }
+  return base;
+};
+
+const validatePropValues = (base, prop) => {
+  const values = {};
+  if (Array.isArray(prop)) {
+    Object.keys(base).forEach((breakpoint, i) => {
+      values[breakpoint] = Number(prop[i]);
+    });
+    return values;
+  }
+  return prop;
+};
+
 export const style = ({ ownerState, theme }) => {
   let styles = {
-    display: 'grid',
-    gridAutoRows: 0,
-    padding: 0,
-    overflow: 'auto',
-    width: '100%',
-    rowGap: 2,
+    display: 'flex',
+    flexFlow: 'column wrap',
+    alignContent: 'space-between',
     boxSizing: 'border-box',
+    '& *': {
+      boxSizing: 'border-box',
+    },
   };
 
-  const base = {};
-  Object.keys(theme.breakpoints.values).forEach((breakpoint) => {
-    if (ownerState.spacing[breakpoint] != null) {
-      base[breakpoint] = true;
-    }
+  const spacingBreakpointsBase = computeBreakpointsBase(theme.breakpoints, ownerState.spacing);
+  const spacingValues = resolveBreakpointValues({
+    values: validatePropValues(spacingBreakpointsBase, ownerState.spacing),
+    base: spacingBreakpointsBase,
   });
 
-  const spacingValues = resolveBreakpointValues({ values: ownerState.spacing, base });
   const transformer = createUnarySpacing(theme);
   const spacingStyleFromPropValue = (propValue) => {
+    propValue = Number(propValue);
+    const spacing = Number(getValue(transformer, propValue).replace('px', ''));
     return {
-      columnGap: getValue(transformer, propValue),
+      margin: -(spacing / 2),
+      '& *': {
```

Comment:
I'm not sure if I understand it correctly. Is this style applied to the direct children of the Masonry? If so, it should be `& > *`, so we don't affect nested elements.

Reply:
You're right! Thanks!